text/0000-collection-transmute.md

```diff
@@ -0,0 +1,152 @@
+    ///
+    /// ```
+    /// let x = vec![0u32; 5];
+    /// let y = unsafe { x.transmute::<[u8; 4]>() };
+    /// assert_eq!(5, y.len());
+    /// assert_eq!([0, 0, 0, 0], y[0]);
+    /// ```
+    ///
+    /// [`mem::transmute`]: ../../std/mem/fn.transmute.html
+    unsafe fn transmute<I>(self) -> Vec<I> {
+        ..
+    }
+```
+
+This would mean our example above would become:
+
+```rust
+let x = vec![0i32; 2];
+let y = unsafe { x.transmute::<[u8; 4]>() };
+```
+
+The documentation of `mem::transmute` should link to the new methods.
+
+A clippy lint can catch offending calls to `mem::transmute` and suggest using 
+the inherent `transmute` method where applicable.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The implementation for `Vec` copies the above solution. The methods for 
+`VecDeque` and `BinaryHeap` use the `Vec` method on their data.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Adding a new method to `std` increases code size and needs to be maintained. 
+However, this seems to be a minor inconvenience when compared to the safety 
+benefit.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+As explained above, the method is useful, yet non-obvious. There were multiple 
+wrong implementation attempts before that were not chosen for being unsound.
+
+We could do nothing, but this means users wanting to transmute the items of a 
+`Vec` will be left without an obvious solution which will lead to unsound code 
+if they get it wrong.
+
+We could document the correct solution instead of putting it into `std`. This 
+would lead to worse ergonomics.
+
+We could create a trait to hold the `transmute` method. This would allow more 
+generic usage, but might lead to worse ergonomics due to type inference 
+uncertainty.
+
+It would even be possible to provide a default implementation using 
+`mem::transmute`, but having a default implementation that might be unsound for 
+some types is a footgun waiting to happen.
+
+# Prior art
+[prior-art]: #prior-art
+
+`mem::transmute` offers the same functionality for many other types. We have 
+added similar methods to different types where useful, see the various 
+iterator-like methods in `std`. @Shnatsel and @danielhenrymantilla came up 
+with the solution together in a 
+[clippy issue](https://github.com/rust-lang/rust-clippy/issues/4484).
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- are there other types that would benefit from such a method? What about 
+`HashSet`, `HashMap` and `LinkedList`?
+- would it make sense to add implementations to types where `mem::transmute` is 
+acceptable, to steer people away from the latter?
+- this RFC does not deal with collection types in crates such as 
+[`SmallVec`](https://docs.rs/smallvec), though it is likely an implementation 
+in `std` might motivate the maintainers to include similar methods.
```

Comment:
FWIW I don't think this is an unresolved question. It is kind of up to those crates to keep up with the libstd collection API. It is normal that they will lag a bit behind, but for the APIs proposed here keeping up is only one PR away.